Expose GRPC KeepAlive parameters to triton client

rmccorm4 opened the pull request:
- [x] Python client library
- [x] simple_grpc_infer_client.py
- [x] C++ client library
- [x] update copyright headers
- [x] document parameters

tanmayv25 commented on src/python/library/tritonclient/grpc/__init__.py:
```diff
@@ -148,12 +148,25 @@ def __init__(self,
                  private_key=None,
-                 certificate_chain=None):
+                 certificate_chain=None,
+                 grpc_arg_keepalive_time_ms=2**31-1,
```
You need to document these parameters above.

tanmayv25 commented on src/python/examples/simple_grpc_infer_client.py:
```diff
@@ -96,7 +96,12 @@
             private_key=FLAGS.private_key,
-            certificate_chain=FLAGS.certificate_chain)
+            certificate_chain=FLAGS.certificate_chain,
+            grpc_arg_keepalive_time_ms=2**31-1,
```
Expose these parameters as the script arguments. Ensure you document the whole thing well. 

These defaults can be easily set in the parser.  See L36-L89.

tanmayv25 commented:
You need to update the copyrights with the current format.

rmccorm4 commented:
GRPC Core will catch and ignore bad values for channel arguments:
```
root@d641fb80ffa3:/mnt/jira/2582-grpc-keepalive/client/src/python/examples# python3 simple_grpc_infer_client.py --grpc-keepalive-time -123 --grpc-keepalive-timeout -456 --grpc-keepalive-permit-without-calls --grpc-http2-max-pi
ngs-without-data -1
E0720 21:41:49.053672236    1470 channel_args.cc:256]        grpc.keepalive_time_ms ignored: it must be >= 1
E0720 21:41:49.054238920    1470 channel_args.cc:256]        grpc.keepalive_time_ms ignored: it must be >= 1
E0720 21:41:49.054250420    1470 channel_args.cc:256]        grpc.keepalive_timeout_ms ignored: it must be >= 1
E0720 21:41:49.054357369    1470 channel_args.cc:256]        grpc.keepalive_time_ms ignored: it must be >= 1
E0720 21:41:49.054365526    1470 channel_args.cc:256]        grpc.keepalive_timeout_ms ignored: it must be >= 0
E0720 21:41:49.054371506    1470 channel_args.cc:256]        grpc.http2.max_pings_without_data ignored: it must be >= 0
...
14 - 1 = 13
15 + 1 = 16
15 - 1 = 14
PASS: infer
```

New help string for `simple_grpc_infer_client.py`:
```
root@d641fb80ffa3:/mnt/jira/2582-grpc-keepalive/client/src/python/examples# python3 simple_grpc_infer_client.py -h
usage: simple_grpc_infer_client.py [-h] [-v] [-u URL] [-s] [-t CLIENT_TIMEOUT] [-r ROOT_CERTIFICATES]
                                   [-p PRIVATE_KEY] [-x CERTIFICATE_CHAIN] [-C GRPC_COMPRESSION_ALGORITHM]
                                   [--grpc-keepalive-time GRPC_KEEPALIVE_TIME]
                                   [--grpc-keepalive-timeout GRPC_KEEPALIVE_TIMEOUT]
                                   [--grpc-keepalive-permit-without-calls]
                                   [--grpc-http2-max-pings-without-data GRPC_HTTP2_MAX_PINGS_WITHOUT_DATA]

optional arguments:
...
  --grpc-keepalive-time GRPC_KEEPALIVE_TIME
                        The period (in milliseconds) after which a keepalive ping is sent on the transport.
                        Default is 2**31-1 (INT_MAX: disabled).
  --grpc-keepalive-timeout GRPC_KEEPALIVE_TIMEOUT
                        The period (in milliseconds) the sender of the keepalive ping waits for an
                        acknowledgement. If it does not receive an acknowledgment within this time, it will
                        close the connection. Default is 20000 (20 seconds).
  --grpc-keepalive-permit-without-calls
                        Allows keepalive pings to be sent even if there are no calls in flight. Default is
                        False.
  --grpc-http2-max-pings-without-data GRPC_HTTP2_MAX_PINGS_WITHOUT_DATA
                        The maximum number of pings that can be sent when there is no data/header frame to be
                        sent. gRPC Core will not continue sending pings if we run over the limit. Setting it to
                        0 allows sending pings without such a restriction. Default is 2.
```

rmccorm4 commented:
@tanmayv25 

1. Do we want to update a C++ example like `simple_grpc_infer_client.cc` to use these params too, or is python client example enough?
2. Am I missing any specific kinds of comments for auto-generated docs [here](https://github.com/triton-inference-server/client/blob/3da940e4ac163b4ed3b57a4f97aa04844ab25372/src/c%2B%2B/library/grpc_client.h#L60-L71) or other places?

tanmayv25 commented:
> Do we want to update a C++ example like simple_grpc_infer_client.cc to use these params too, or is python client example enough?

We definitely need a C++ example to demonstrate the use of keep alive parameters for C++ library. I would suggest having another example and call it simple_grpc_keepalive_client.cc instead of polluting the current `simple_grpc_infer_client.cc` example with these options. 

> Am I missing any specific kinds of comments for auto-generated docs here or other places?

Only the python side is auto-generated docs which again was obsoleted with our move to markdowns in our project. That said we must still document everything in the user side interface to our libraries which includes your attached links. I will submit a detailed review later.

tanmayv25 commented on src/c++/library/grpc_client.h:
```diff
@@ -57,6 +57,19 @@ struct SslOptions {
+        keepalive_permit_without_calls(false), http2_max_pings_without_data(2)
+  {
+  }
+  int keepalive_time_ms;
```
We must definitely document these parameters here.

tanmayv25 commented on src/python/examples/simple_grpc_infer_client.py:
```diff
@@ -85,18 +85,68 @@
+        'Default is None.'
+    )
+    # GRPC KeepAlive: https://github.com/grpc/grpc/blob/master/doc/keepalive.md
+    parser.add_argument(
```
I think will be better to have another example to demonstrate keep alive settings. The simple_grpc_infer_client.py should be simple. These many new options pollute our simple example.

Let's capture this in `simple_grpc_keepalive_client.py` ?

tanmayv25 commented on src/c++/library/grpc_client.h:
```diff
@@ -89,7 +102,8 @@ class InferenceServerGrpcClient : public InferenceServerClient {
       const std::string& server_url, bool verbose = false, bool use_ssl = false,
-      const SslOptions& ssl_options = SslOptions());
+      const SslOptions& ssl_options = SslOptions(),
+      const KeepAliveOptions& keepalive_options = KeepAliveOptions());
```
Document `keepalive_options` as one of the parameters above after L100.

rmccorm4 commented:
1. Addressed documentation comments above
2. Added `simple_grpc_keepalive_client.py` and `simple_grpc_keepalive_client.cc`

c++ help string
```
root@RMCCORMICK:/mnt/jira/2582-grpc-keepalive/client/build/install/bin# ./simple_grpc_keepalive_client -h
./simple_grpc_keepalive_client: invalid option -- 'h'
Usage: ./simple_grpc_keepalive_client [options]
        -v
        -m <model name>
        -u <URL for inference service>
        -t <client timeout in microseconds>
        -H <HTTP header>
        --grpc-keepalive-time <milliseconds>
        --grpc-keepalive-timeout <milliseconds>
        --grpc-keepalive-permit-without-calls
        --grpc-http2-max-pings-without-data <number of pings>

For -H, header must be 'Header:Value'. May be given multiple times.
```
c++ test
```
root@RMCCORMICK:/mnt/jira/2582-grpc-keepalive/client/build/install/bin# ./simple_grpc_keepalive_client  --grpc-keepalive-time 1234 --grpc-keepalive-timeout 5678 --grpc-keepalive-permit-without-calls --grpc-http2-max-pings-without-data 42
...
PASS : Infer
```
python help string
```
root@RMCCORMICK:/mnt/jira/2582-grpc-keepalive/client/src/python/examples# python3 simple_grpc_keepalive_client.py -h
usage: simple_grpc_keepalive_client.py [-h] [-v] [-u URL] [--grpc-keepalive-time GRPC_KEEPALIVE_TIME] [--grpc-keepalive-timeout GRPC_KEEPALIVE_TIMEOUT] [--grpc-keepalive-permit-without-calls] [--grpc-http2-max-pings-without-data GRPC_HTTP2_MAX_PINGS_WITHOUT_DATA]

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         Enable verbose output
  -u URL, --url URL     Inference server URL. Default is localhost:8001.
  --grpc-keepalive-time GRPC_KEEPALIVE_TIME
                        The period (in milliseconds) after which a keepalive ping is sent on the transport. Default is 2**31-1 (INT_MAX: disabled).
  --grpc-keepalive-timeout GRPC_KEEPALIVE_TIMEOUT
                        The period (in milliseconds) the sender of the keepalive ping waits for an acknowledgement. If it does not receive an acknowledgment within this time, it will close the connection. Default is 20000 (20 seconds).
  --grpc-keepalive-permit-without-calls
                        Allows keepalive pings to be sent even if there are no calls in flight. Default is False.
  --grpc-http2-max-pings-without-data GRPC_HTTP2_MAX_PINGS_WITHOUT_DATA
                        The maximum number of pings that can be sent when there is no data/header frame to be sent. gRPC Core will not continue sending pings if we run over the limit. Setting it to 0 allows sending pings without such a restriction. Default is 2.
```
python test
```
root@RMCCORMICK:/mnt/jira/2582-grpc-keepalive/client/src/python/examples# python3 simple_grpc_keepalive_client.py --grpc-keepalive-time 1234 --grpc-keepalive-timeout 5678 --grpc-keepalive-permit-without-calls --grpc-http2-max-pings-without-data 42
...
PASS: infer
```

CoderHam commented on src/c++/examples/simple_grpc_keepalive_client.cc:
```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2020-2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Fix copyright year to 2021 only

CoderHam commented on src/c++/library/grpc_client.h:
```diff
@@ -1,4 +1,5 @@
-// Copyright (c) 2020-2021, NVIDIA CORPORATION. All rights reserved.
+// Copyright (c) 2020-2021, NVIDIA CORPORATION & AFFILIATES. All rights
```
Remove (c) to avoid overflowing to next line

CoderHam commented on src/c++/library/grpc_client.h:
```diff
@@ -57,6 +58,29 @@ struct SslOptions {
+        keepalive_permit_without_calls(false), http2_max_pings_without_data(2)
+  {
+  }
+  // The period (in milliseconds) after which a keepalive ping is sent on the
```
on ->over

rmccorm4 replied:
These descriptions are taken directly from https://github.com/grpc/grpc/blob/master/doc/keepalive.md#keepalive-user-guide-for-grpc-core-and-dependents

tanmayv25 commented on src/python/examples/simple_grpc_keepalive_client.py:
```diff
@@ -0,0 +1,159 @@
+        outputs=outputs,
+        headers={'test': '1'})
+
+    statistics = triton_client.get_inference_statistics(model_name=model_name)
```
You can remove the statitistics call.

tanmayv25 commented on src/python/examples/simple_grpc_keepalive_client.py:
```diff
@@ -0,0 +1,159 @@
+            print("sync infer error: incorrect difference")
+            sys.exit(1)
+
+    print('PASS: infer')
```
PASS: keepalive ??

tanmayv25 commented on src/c++/examples/simple_grpc_keepalive_client.cc:
```diff
@@ -0,0 +1,313 @@
+  // Get full response
+  std::cout << results_ptr->DebugString() << std::endl;
+
+  tc::InferStat infer_stat;
```
You can remove L292-307 from here.

tanmayv25 commented on src/c++/examples/simple_grpc_keepalive_client.cc:
```diff
@@ -0,0 +1,313 @@
+  std::cout << model_stat.DebugString() << std::endl;
+
+
+  std::cout << "PASS : Infer" << std::endl;
```
`Pass: Keepalive` ?
Even though it is not specifically testing keepalive behavior.. You can use these examples for some level of sanity testing of the feature.. Same for python client example below.
